src/app/gui/queryresults/queryresultsservice.js

```diff
@@ -1447,342 +1816,428 @@ proto.downloadFeatures = async function(type, layer, features=[], action, index)
-        if (type === 'polygon'){
-          // id type polygon add paramateres to api download
+      // choose between only feature attribute or also polygon attibute
+      download: (type) => {
+        if ('polygon' === type) { // id type polygon add paramateres to api download
           data.sbp_qgs_layer_id = polygonLayer.getId();
           data.sbp_fid = fid;
-        } else {
-          // force to remove
+        } else {                  // force to remove
           delete data.sbp_fid;
           delete data.sbp_qgs_layer_id;
         }
         runDownload(true)
       }
     };
-    if (features.length === 1) {
-      if (downloadsactions === undefined) {
-        action.state.toggled[index] = true;
-      }
-      this.state.actiontools[QueryPolygonCsvAttributesComponent.name] = this.state.actiontools[layerId] || {};
-      this.state.actiontools[QueryPolygonCsvAttributesComponent.name][layerId] = config;
-      this.setCurrentActionLayerFeatureTool({
-        layer,
-        index,
-        action,
-        component: QueryPolygonCsvAttributesComponent
-      });
-    } else {
-      if (downloadsactions === undefined) {
-        layer[type].active = !layer[type].active;
-        if (layer[type].active) {
-          this.setLayerActionTool({
-            layer,
-            component: QueryPolygonCsvAttributesComponent,
-            config
-          });
-        } else this.setLayerActionTool({
-          layer
-        })
-      } else {
-        this.setLayerActionTool({
-          layer,
-          component: QueryPolygonCsvAttributesComponent,
-          config
-        });
-      }
+
+    /** @FIXME add description */
+    if (1 === features.length && undefined === downloadsactions) {
+      action.state.toggled[index] = true;
+    }
+
+    /** @FIXME add description */
+    if(1 === features.length) {
+      this.state.actiontools[QueryPolygonCsvAttributesComponent.name] = this.state.actiontools[layer.id] || {};
+      this.state.actiontools[QueryPolygonCsvAttributesComponent.name][layer.id] = config;
+      this.setCurrentActionLayerFeatureTool({ layer, index, action, component: QueryPolygonCsvAttributesComponent });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined === downloadsactions) {
+      layer[type].active = !layer[type].active;
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined === downloadsactions && layer[type].active) {
+        this.setLayerActionTool({ layer, component: QueryPolygonCsvAttributesComponent, config });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined === downloadsactions && !layer[type].active){
+      this.setLayerActionTool({ layer });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined !== downloadsactions) {
+      this.setLayerActionTool({ layer, component: QueryPolygonCsvAttributesComponent, config });
     }
-  } else runDownload();
+
+  }
```

Comment:
within this function `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`, i rewrote several if-else conditions to reduce nesting level, check carefully